staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go

```diff
@@ -21,39 +21,246 @@ import (
-			container := &pod.Spec.Containers[i]
-			if container.Name != containerName {
-				continue
+const (
+	// debug by ephemeral container
+	ephemeral debugStyle = iota
+	// debug by pod copy
+	podCopy
+	// debug node
+	node
+	// unsupported debug methodology
+	unsupported
+)
+
+func getDebugStyle(pod *corev1.Pod, target runtime.Object) (debugStyle, error) {
+	switch target.(type) {
+	case *corev1.Pod:
+		if asserted, ok := target.(*corev1.Pod); ok {
+			if pod != asserted { // comparing addresses
+				return podCopy, nil
 			}
-			container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
-				MountPath: "/host",
-				Name:      volumeName,
-			})
 		}
+		return ephemeral, nil
+	case *corev1.Node:
+		return node, nil
+	}
+	return unsupported, fmt.Errorf("objects of type %T are not supported", target)
+}
 
-		pod.Spec.HostIPC = true
-		pod.Spec.HostNetwork = true
-		pod.Spec.HostPID = true
-		return nil
-	default:
-		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
+func (p *generalProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	style, err := getDebugStyle(pod, target)
+	if err != nil {
+		return fmt.Errorf("general profile: %s", err)
+	}
+
+	switch style {
+	case node:
+		MountRootPartition(pod, containerName)
+		ClearSecurityContext(pod, containerName, podutils.Containers)
+		UseHostNamespaces(pod)
+
+	case podCopy:
+		RemoveLabelsAndProbes(pod)
+		AllowProcessTracing(pod, containerName, podutils.Containers)
+		ShareProcessNamespace(pod)
+
+	case ephemeral:
+		AllowProcessTracing(pod, containerName, podutils.EphemeralContainers)
+	}
+
+	return nil
+}
+
+func (p *baselineProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	style, err := getDebugStyle(pod, target)
+	if err != nil {
+		return fmt.Errorf("baseline profile: %s", err)
+	}
+
+	ClearSecurityContext(pod, containerName, podutils.Containers|podutils.EphemeralContainers)
+
+	switch style {
+	case podCopy:
+		RemoveLabelsAndProbes(pod)
+		ShareProcessNamespace(pod)
+
+	case ephemeral, node:
+		// no additional modifications needed
 	}
+
+	return nil
+}
+
+func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	style, err := getDebugStyle(pod, target)
+	if err != nil {
+		return fmt.Errorf("restricted profile: %s", err)
+	}
+
+	DisallowRoot(pod, containerName, podutils.Containers|podutils.EphemeralContainers)
+	DropCapabilities(pod, containerName, podutils.Containers|podutils.EphemeralContainers)
+
+	switch style {
+	case node:
+		ClearSecurityContext(pod, containerName, podutils.Containers)
+
+	case podCopy:
+		ShareProcessNamespace(pod)
+
+	case ephemeral:
+		// no additional modifications needed
+	}
+
+	return nil
+}
+
+// RemoveLabelsAndProbes removes labels from the pod and remove probes
+// from all containers of the pod.
+func RemoveLabelsAndProbes(p *corev1.Pod) {
```

Comment:
nit: maybe we can unexport all these helper functions, if there isn't any explicit reason

Reply:
+1

Reply:
sounds good - done in 98f8ff5a3ef6516128987798111313f32ffc7b41